designs/2019-config-simplification/README.md

```diff
@@ -0,0 +1,627 @@
+
+**Note:** This would allow us to stop shipping environments in ESLint. We could just tell people to use `globals` in their config and allow them to specify which version of `globals` they want to use.
+
+#### Extending Another Config
+
+Extending another config works the same as in `.eslintrc` except users pass entire config objects rather than the name of a config. That can easily be done in the following manner:
+
+```js
+exports.config = {
+    files: ["*.js"],
+    extends: require("eslint-config-standard"),
+    rules: {
+        semi: ["error", "always"]
+    }
+};
+```
+
+This config extends `eslint-config-standard` by assigning it to `extends`. You can also use an array for `extends` to extend from multiple configs:
+
+```js
+exports.config = {
+    files: ["*.js"],
+    extends: [
+        require("eslint-config-standard"),
+        require("@me/eslint-config")
+    ],
+    rules: {
+        semi: ["error", "always"]
+    }
+};
+```
+
+#### Extending From `eslint:recommended` and `eslint:all`
+
+Both `eslint:recommended` and `eslint:all` can be represented as strings in the `extends` key. For example:
+
+```js
+exports.config = {
+    files: ["*.js"],
+    extends: [
+        "eslint:recommended",
+        require("eslint-config-standard"),
+        require("@me/eslint-config")
+    ],
+    rules: {
+        semi: ["error", "always"]
+    }
+};
+```
+
+This config first extends `eslint:recommended` and then continues on to extend other configs.
+
+#### Overriding Configuration Based on File Patterns
+
+Whereas `.eslintrc` had an `overrides` key that made a hierarchical structure, the `eslint.config.js` file does not have any such hierarchy. Instead, users can return an array of configs that should be used. For example, consider this `.eslintrc` config:
+
+```yaml
+plugins: ["react"]
+rules:
+   react/jsx-uses-react: error
+    semi: error
+
+overrides:
+    - files: "*.md"
+      plugins: ["markdown"],
+      processor: "markdown/markdown"
+```
+
+This can be written in `eslint.config.js` as an array of two configs:
+
+```js
+exports.config = [
+    {
+        files: "*.js",
+        ruledefs: {
+            react: require("eslint-plugin-react").rules,
+        },
+        rules: {
+            "react/jsx-uses-react": "error",
+            semi: "error"
+        }
+    },
+    {
+        files: "*.md",
+        processor: require("eslint-plugin-markdown").processors.markdown
+    }
+];
+```
+
+When ESLint uses this config, it will check each `files` pattern to determine which configs apply. Any config with a `files` pattern matching the file to lint will be extracted and used (if multiple configs match, then those configs are merged to determine the final config to use). In this way, returning an array acts exactly the same as the array in `overrides`.
+
+If a config in the config array does not contain `files` or `ignores`, then that config applies to all files. For example:
```

Comment:
Yes, the config returned by `exports.config` must have a `files` property in order to match against the files that are passed on the command line. However, any single object inside of a config array need not have `files` so that it applies to everything. I'll add some text to make that clear.